libdrgn/kallsyms.c

```diff
@@ -0,0 +1,625 @@
+	kr->token_index = malloc(token_index_size);
+	if (!kr->token_index)
+		return &drgn_enomem;
+	err = drgn_program_read_memory(prog, kr->token_index,
+				       loc->kallsyms_token_index,
+				       token_index_size, false);
+	if (err)
+		return err;
+	if (bswap) {
+		for (size_t i = 0; i < kr->num_syms; i++) {
+			kr->token_index[i] = bswap_16(kr->token_index[i]);
+		}
+	}
+
+	/*
+	 * Find the end of the last token, so we get the overall length of
+	 * token_table. Then copy the token_table into host memory.
+	 */
+	last_token = loc->kallsyms_token_table + kr->token_index[UINT8_MAX];
+	do {
+		err = drgn_program_read_u8(prog, last_token, false,
+					   (uint8_t *)&data);
+		if (err)
+			return err;
+
+		last_token++;
+	} while (data);
+	token_table_size = last_token - loc->kallsyms_token_table + 1;
+	kr->token_table = malloc(token_table_size);
+	if (!kr->token_table)
+		return &drgn_enomem;
+	err = drgn_program_read_memory(prog, kr->token_table,
+				       loc->kallsyms_token_table,
+				       token_table_size, false);
+	if (err)
+		return err;
+
+	/* Now find the end of the names array by skipping through it, then copy
+	 * that into host memory. */
+	names_idx = 0;
+	kr->long_names = guess_long_names(prog);
+	for (size_t i = 0; i < kr->num_syms; i++) {
+		err = drgn_program_read_u8(prog,
+					   loc->kallsyms_names + names_idx,
+					   false, &len_u8);
+		if (err)
+			return err;
+		len = len_u8;
+		if ((len & 0x80) && kr->long_names) {
+			err = drgn_program_read_u8(prog,
+						loc->kallsyms_names + names_idx + 1,
+						false, &len_u8);
+			if (err)
+				return err;
+			len = (len & 0x7F) | (len_u8 << 7);
+			names_idx++;
+		}
+		names_idx += len + 1;
+	}
+	kr->names = malloc(names_idx);
+	if (!kr->names)
+		return &drgn_enomem;
+	err = drgn_program_read_memory(prog, kr->names,
+				       loc->kallsyms_names,
+				       names_idx, false);
+	if (err)
+		return err;
+
+	return NULL;
+}
+
+/**
+ * Extract the symbol name and type
+ * @param kr Registry containing kallsyms data
+ * @param offset Starting index within names array for this symbol
+ * @param sb Buffer to write output symbol to
+ * @param[out] kind_ret Where to write the symbol kind data
+ * @returns The offset of the next symbol in the names array
+ */
+static unsigned int
+kallsyms_expand_symbol(struct kallsyms_reader *kr, unsigned int offset,
+		       struct string_builder *sb, char *kind_ret)
+{
+	uint8_t *data = &kr->names[offset];
+	unsigned int len = *data;
+	bool skipped_first = false;
+
+	if ((len & 0x80) && kr->long_names) {
+		data++;
+		offset++;
+		len = (0x7F & len) | (*data << 7);
+	}
+
+	offset += len + 1;
+	data += 1;
+	while (len) {
+		char *token_ptr = &kr->token_table[kr->token_index[*data]];
+		while (*token_ptr) {
+			if (skipped_first) {
+				if (!string_builder_appendc(sb, *token_ptr)) {
+					string_builder_null_terminate(sb);
+					return offset;
+				}
+			} else {
+				if (kind_ret)
+					*kind_ret = *token_ptr;
```

Comment:
Can we return without ever setting this, leaving the caller with uninitialized garbage?

Reply:
Yes, I think that's worth guarding against. I think the easiest place to do it is actually in `drgn_load_builtin_kallsyms()`, where we can simply check whether the string builder length is zero and raise an error.

There's a lot of stuff in the kallsyms tables that, if fuzzed, could cause these parsers to behave weirdly or maybe crash. I'm not certain how paranoid we want to be here. At this point, if the correct set of symbols are provided, with real values, and we've managed to read the token and name table successfully (without faulting or mallocing a buffer that's way too large), then it seems likely that the data will match the kallsyms format. So in practical terms, these sorts of things shouldn't happen. What are your opinions here? I can add paranoia to this, but it does feel like it will take rather unclear code and make it even worse...